### PR TITLE
Parallelize priority and fit.

### PR DESCRIPTION
**Pull request by brendandburns**
@davidopp @lavalamp 

Ref #17389

### CODE REVIEW

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 1]
```diff
@@ -113,30 +133,31 @@ func findNodesThatFit(pod *api.Pod, podLister algorithm.PodLister, predicateFunc
...
-				if _, found := failedPredicateMap[node.Name]; !found {
-					failedPredicateMap[node.Name] = sets.String{}
-				}
-				if predicates.FailedResourceType != "" {
-					failedPredicateMap[node.Name].Insert(predicates.FailedResourceType)
-					break
-				}
-				failedPredicateMap[node.Name].Insert(name)
-				break
+			lock.Lock()
```
Comment (lavalamp) [inline plugin/pkg/scheduler/generic_scheduler.go]: Sending results back on a channel probably makes more sense than using a lock here?

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 2]
```diff
@@ -157,20 +178,37 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
...
+			if weight == 0 {
+				return
+			}
+			priorityFunc := priorityConfig.Function
+			prioritizedList, err := priorityFunc(pod, podLister, nodeLister)
+			if err != nil {
+				returnErr = err
+			}
+
+			lock.Lock()
```
Comment (lavalamp) [inline plugin/pkg/scheduler/generic_scheduler.go]: Same here.

**Comment by lavalamp** [comment]
Can you benchmark this or otherwise measure it?

Possibly also makes sense to run priority & fit functions themselves in parallel.

**Comment by davidopp** [comment]
@erictune who was also interested in this

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra @smarterclayton

**Comment by ncdc** [comment]
cc @timothysc @jeremyeder

**Comment by davido2** [comment]
@lavalamp It's not obvious to me that it is necessary to run the priority
and predicate functions in parallel. I think it is very dependent on the
workload and machines' labels and fullness. The less complexity we can add,
the better. (I realize that what you're describing isn't very complex.)

We can benchmark against density and scalability tests in Kubenark  with
large clusters (there is already a Jenkins for this).

**Comment by brendandburns** [comment]
Yeah, I'm leaning towards agreeing.  My initial benchmarking is indicating that this doesn't help very much.  (on the up side, I'm getting together decent micro benchmarks ;)

I'll try to close this out one way or the other this evening.

**Comment by davidopp** [comment]
To be clear, I wasn't suggesting this PR is a bad idea, just that exeltdnding it to also run feasibility and scoring in parallel may not be a big win.

I'd actually be surprised if this PR isn't helpful for large clusters in scenarios where the predicate or priority functions have to do a lot of work. (though it's possible we don't yet have such functions, but may in the future)

**Comment by brendandburns** [comment]
So concretely:

Currently, running a single predicate across 1000 nodes x 30000 pods takes 0.16 _milliseconds_

2x more time is spent in `predicates.MapPodsToMachines` than in the predicate function.

I think our (current predicates at least) are simply too fast for this to matter that much.

I will polish up the micro benchmark and send it for review.  We can use that to test subsequent ideas.

**Comment by brendandburns** [comment]
Ok PR with the benchmark is here:

https://github.com/kubernetes/kubernetes/pull/17415

Results with this PR:
```sh
Running tests for APIVersion: v1,extensions/v1beta1,metrics/v1alpha1 with etcdPrefix: registry
+++ [1117 20:35:17] Running tests without code coverage
PASS
BenchmarkPredicates	     100	  22565348 ns/op	 4825257 B/op	   41194 allocs/op
BenchmarkPriorities	      50	  33762478 ns/op	 9431946 B/op	   97342 allocs/op
ok  	k8s.io/kubernetes/plugin/pkg/scheduler	4.300s
Running tests for APIVersion: v1,extensions/v1beta1,metrics/v1alpha1 with etcdPrefix: kubernetes.io/registry
+++ [1117 20:35:33] Running tests without code coverage
PASS
BenchmarkPredicates	      50	  22943546 ns/op	 4824229 B/op	   41191 allocs/op
BenchmarkPriorities	      50	  33466442 ns/op	 9432890 B/op	   97344 allocs/op
ok  	k8s.io/kubernetes/plugin/pkg/scheduler	3.218s
```

Results w/o this PR:
```sh
+++ [1117 20:34:38] Running tests without code coverage
PASS
BenchmarkPredicates	     100	  18678852 ns/op	 4745046 B/op	   40184 allocs/op
BenchmarkPriorities	      50	  33037231 ns/op	 9432576 B/op	   97334 allocs/op
ok  	k8s.io/kubernetes/plugin/pkg/scheduler	3.861s
Running tests for APIVersion: v1,extensions/v1beta1,metrics/v1alpha1 with etcdPrefix: kubernetes.io/registry
+++ [1117 20:34:54] Running tests without code coverage
PASS
BenchmarkPredicates	     100	  18493052 ns/op	 4744469 B/op	   40183 allocs/op
BenchmarkPriorities	      50	  30896857 ns/op	 9433206 B/op	   97336 allocs/op
```

**Comment by lavalamp** [comment]
Thanks, this matters even less than I expected :)

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 3]
```diff
@@ -113,30 +133,31 @@ func findNodesThatFit(pod *api.Pod, podLister algorithm.PodLister, predicateFunc
...
+	wait.Add(len(nodes.Items))
+	lock := sync.Mutex{}
+	var returnErr error
+	for ix := range nodes.Items {
+		go func(node *api.Node) {
+			defer wait.Done()
+			fits, failedPredicates, err := testNodeFit(pod, node, machineToPods, predicateFuncs)
 			if err != nil {
-				return api.NodeList{}, FailedPredicateMap{}, err
+				// Yes this could clobber other errors, but I don't care
```
Comment (pmorie) [inline plugin/pkg/scheduler/generic_scheduler.go]: Tell us how you really feel

**Pull request closed without merging**